ethers-core/ethers-derive-eip712/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+//! determine if there is a nested eip712 struct. However, this work is not yet complete.
+//!
+use std::convert::TryFrom;
+
+use ethers_core::types::transaction::eip712;
+use proc_macro::TokenStream;
+use quote::quote;
+
+#[proc_macro_derive(Eip712, attributes(eip712))]
+pub fn eip_712_derive(input: TokenStream) -> TokenStream {
+    let ast = syn::parse(input).expect("failed to parse token stream for Eip712 derived struct");
+
+    impl_eip_712_macro(&ast)
+}
+
+// Main implementation macro, used to compute static values and define
+// method for encoding the final eip712 payload;
+fn impl_eip_712_macro(ast: &syn::DeriveInput) -> TokenStream {
+    // Primary type should match the type in the ethereum verifying contract;
+    let primary_type = &ast.ident;
+
+    // Computer domain separator
+    let domain = match eip712::EIP712Domain::try_from(ast) {
+        Ok(attributes) => attributes,
+        Err(e) => return TokenStream::from(e),
+    };
+
+    let domain_separator = hex::encode(domain.separator());
+
+    // Must parse the AST at compile time.
+    let parsed_fields = match eip712::parse_fields(ast) {
+        Ok(fields) => fields,
+        Err(e) => return TokenStream::from(e),
+    };
+
+    // Compute the type hash for the derived struct using the parsed fields from above;
+    let type_hash = hex::encode(eip712::make_type_hash(
+        primary_type.clone().to_string(),
+        &parsed_fields,
+    ));
+
+    let implementation = quote! {
+        impl Eip712 for #primary_type {
+            type Error = ethers_core::types::transaction::eip712::Eip712Error;
+
+            fn type_hash() -> Result<[u8; 32], Self::Error> {
+                use std::convert::TryFrom;
+                let decoded = hex::decode(#type_hash.to_string())?;
+                let byte_array: [u8; 32] = <[u8; 32]>::try_from(&decoded[..])?;
+                Ok(byte_array)
+            }
+
+            fn domain_separator() -> Result<[u8; 32], Self::Error> {
+                use std::convert::TryFrom;
+                let decoded = hex::decode(#domain_separator.to_string())?;
+                let byte_array: [u8; 32] = <[u8; 32]>::try_from(&decoded[..])?;
+                Ok(byte_array)
+            }
+
+            fn struct_hash(self) -> Result<[u8; 32], Self::Error> {
+                use ethers_core::abi::Tokenizable;
+                let mut items = vec![ethers_core::abi::Token::Uint(
+                    ethers_core::types::U256::from(&Self::type_hash()?[..]),
+                )];
+
+                if let ethers_core::abi::Token::Tuple(tokens) = self.clone().into_token() {
+                    for token in tokens {
+                        match &token {
+                            ethers_core::abi::Token::Tuple(t) => {
+                                // TODO: check for nested Eip712 Type;
+                                // Challenge is determining the type hash
```

Comment:
should this fail for the time being?

Reply:
fails with error now, but should be unreachable in practice. Will think about a solution for the nested struct, but it might require some re-working to get it done. Hopefully this is a good start.